url73e3057d61_empeeoh_BACnet4J_tgz-pJ8-com_serotonin_bacnet4j_test_OverflowTestJ8/src/com/serotonin/bacnet4j/transport/Transport.java

```diff
@@ -0,0 +1,588 @@
+                        network.getMaxApduLength(), key.getInvokeId(), 0, segWindow, service.getChoiceId(),
+                        (ByteQueue) null);
+                ack = sendSegmented(key, maxAPDULengthAccepted, maxServiceData, serviceData, template);
+                if (ack == null)
+                    // Go to the waiting room to wait for a response.
+                    ack = waitForAck(key, timeout, true);
+            }
+            finally {
+                waitingRoom.leave(key);
+            }
+        }
+        else {
+            // We can send the whole APDU in one shot.
+            final WaitingRoomKey key = waitingRoom.enterClient(address, ls);
+            final ConfirmedRequest apdu = new ConfirmedRequest(false, false, true, 
+            							MAX_SEGMENTS, network.getMaxApduLength(),
+            							key.getInvokeId(), (byte) 0, 0, 
+            							service.getChoiceId(), serviceData);
+            try {
+                ack = sendSegments(key, timeout, new APDU[] { apdu });
+            }
+            finally {
+                waitingRoom.leave(key);
+            }
+        }
+
+        if (ack instanceof SimpleACK)
+            return null;
+
+        if (ack instanceof ComplexACK)
+            return ((ComplexACK) ack).getService();
+
+        if (ack instanceof Error)
+            throw new ErrorAPDUException((Error) ack);
+
+        if (ack instanceof Reject)
+            throw new RejectAPDUException((Reject) ack);
+
+        if (ack instanceof Abort)
+            throw new AbortAPDUException((Abort) ack);
+
+        throw new BACnetException("Unexpected ack APDU: " + ack);
+    }
+
+    public Address getLocalBroadcastAddress() {
+        return network.getLocalBroadcastAddress();
+    }
+
+    public void sendUnconfirmed(final Address address, 
+    							final OctetString linkService, 
+    							final UnconfirmedRequestService serviceRequest,
+    							final boolean broadcast) throws BACnetException {
+    	if (address == null)
+    		throw new IllegalArgumentException("Address cannot be null");
+    	// Unconfirmed services will never have to be segmented, so just send it.
+    	if (address.equals(linkService)){
+    		network.sendAPDU(address, null, new UnconfirmedRequest(serviceRequest), broadcast);
+    	}else{
+    		network.sendAPDU(address, linkService, new UnconfirmedRequest(serviceRequest), broadcast);   
+    	}
+    }
+
+    //
+    // Receiving messages - messages are initiated externally, and received from the data link. Just call the handler
+    // method in the service.
+    //
+    private AcknowledgementService handleConfirmedRequest(Address from, OctetString linkService, byte invokeId,
+            ConfirmedRequestService service) throws BACnetException {
+        try {
+            return service.handle(localDevice, from, linkService);
+        }
+        catch (NotImplementedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 257.